### ISSUE TITLE
alias_method vs prepend infinite looping

### ISSUE THREAD

**Issue by sribalakumar** [open-telemetry/opentelemetry-ruby#352]
When opentelemetry-ruby instrumentations is run in parallel with other Gems (like [prometheus-exporter](https://github.com/discourse/prometheus_exporter)) which use a different patching technique and patch the same low level adapter method, we observe `SystemStackError (stack level too deep)` error at runtime.

More details related to `SystemStackError` originating because of patching conflicts between `alias_method` vs `Module#prepend` can be read from https://blog.newrelic.com/engineering/ruby-agent-module-prepend-alias-method-chains/

The patching problem between both techniques looks to a common issue among the APM providers community.
    - elastic/apm-agent-ruby#379
    - https://docs.newrelic.com/docs/agents/ruby-agent/troubleshooting/systemstackerror-stack-level-too-deep
    - scoutapp/scout_apm_ruby#309 (data dog integration)
    - DataDog/dd-trace-rb#862

May be we can provide a config toggle flag for the end user to choose which technique they need [similar to New Relic implementation](https://github.com/newrelic/newrelic-ruby-agent/blob/c9203be16af3468f275bbde078dc2805ad878655/lib/new_relic/agent/instrumentation/active_record.rb#L23) ?

New Relic config reference:
https://docs.newrelic.com/docs/agents/ruby-agent/configuration/ruby-agent-configuration
> If `prepend_active_record_instrumentation: true`, uses Module.prepend rather than alias_method for ActiveRecord instrumentation.

**Comment by lloeki** [comment]
I encourage you to read this feedback I made to APM vendors: https://github.com/elastic/apm-agent-ruby/issues/379#issuecomment-717341248

tl;dr: just use prepend.

FYI list of instrumentors now using prepend:

- `scout_apm` >= 2.5.2
- `ddtrace` >= 0.27
- `skylight` >= 5.0.0.beta
- `elastic-apm` >= 4.0 (upcoming)
- `airbrake` >= 11.0.2
- `newrelic_rpm` >= 6.14 (partially: Rack's to_app is still alias method chain'ed, and ActiveRecord is behind `prepend_active_record_instrumentation` that defaults to `false`)

At Sqreen we went the extra mile and detect those to fall back to alias method chain, or warn and bail out when there's a mix, but as of today just using prepend consistently is the solid choice.

**Comment by miquelbar** [comment]
I was being able to execute Opentelemetry with Prometheus exporter using the latest version of Prometheus Exporter. The latest version of Prometheus Exporter seems to work also with prepend: https://github.com/discourse/prometheus_exporter#choosing-the-style-of-method-patching

**Comment by ahayworth** [comment]
We only have one instance of `alias_method` remaining in the codebase at this point. I think we're alright to close this out.
